blog/2025-01-13-typescript-sdk-release.md

```diff
@@ -0,0 +1,375 @@
+
+```typescript
+import { AppBskyEmbedImages } from '@atproto/api'
+
+// Alias, for clarity
+const isImages = AppBskyEmbedImages.isMain
+
+// Get an invalid embed somehow
+const invalidEmbed = {
+  $type: 'app.bsky.embed.images',
+  // notice how the `images` property is missing here
+}
+
+// This predicate function only checks the value of the `$type` property, making the condition "true" here
+if (isImages(invalidEmbed)) {
+  // No TypeScript error, BUT causes a runtime error because there is no "images" property !
+  console.log('First image:', invalidEmbed.images[0])
+}
+```
+
+The root of the issue here is that the `is*` utility methods perform type casting of objects solely based on the value of their `$type` property. There were basically two ways of fixing this issue:
+
+1. Alter the implementation to actually validate the object's structure. This is a non-breaking change that has a negative impact on performance.
+2. Alter the function signature to describe what the function actually does. This is a breaking change because TypeScript would start (rightfully) returning lots of errors in places where these functions are used.
+
+Because this release introduces other breaking changes, and because adapting our own codebase to this change showed it made more sense, we decided to adopt the latter option.
+
+In lots of cases where data needs to be discriminated, this change in the signature of the `is*` function will actually not cause any issue when upgrading the version of the SDK. This is the case for example when working with data obtained from the API. Because an API is a "contract" between a server and a client, the data returned by the server is "guaranteed" to be valid. In these cases, the `is*` utility methods provide a convenient way to discriminate between valid values.
+
+```typescript
+import { AppBskyEmbedImages } from '@atproto/api'
+
+// Aliased for clarity
+const isImages = AppBskyEmbedImages.isMain
+
+// Get a post from the API (the API's contract guarantees the validity of the data)
+declare const post: BlueskyPost
+
+// The `is*` utilities are an efficient way to discriminate **valid** data based on their `$type`
+if (isImages(post.embed)) {
+  // `post.embed` is fully typed as `$Typed<AppBskyEmbedImages.Main>` here !
+}
+```
+
+For other cases, when the data's validity is not known at dev time, we added new `isValid*` utility methods allowing to ensure that a value properly satisfies the interface.
+
+```typescript
+import { AppBskyEmbedImages } from '@atproto/api'
+
+// Aliased for clarity
+const Images = AppBskyEmbedImages.Main
+const isValidImages = AppBskyEmbedImages.isValidMain
+
+// Get an embed with unknown validity somehow
+declare const embed: unknown
+
+// The following condition will be true if, and only if, the value matches the `Image` interface
```

Comment:
Nit: lot of these comments throughout this doc scroll off the screen. Might be good to do a multi-line comment so that readers don't have to scroll horizontally to read the comments.